pkg/operator/ceph/cluster/mon/mon.go

```diff
@@ -1188,6 +1194,116 @@ func (c *Cluster) saveMonConfig() error {
 }
 
 func (c *Cluster) persistExpectedMonDaemons() error {
+	if err := c.persistExpectedMonDaemonsInConfigMap(); err != nil {
+		return errors.Wrap(err, "failed to persist expected mons in ConfigMap")
+	}
+
+	if err := c.persistExpectedMonDaemonsAsEndpointSlice(); err != nil {
+		return errors.Wrap(err, "failed to persist expected mons in EndpointSlice")
+	}
+	return nil
+}
+
+func (c *Cluster) persistExpectedMonDaemonsAsEndpointSlice() error {
+	monitors := c.ClusterInfo.AllMonitors()
+	if len(monitors) == 0 {
+		// Theoretically, we could now go ahead with the normal code path to
+		// delete (both IPv4 and IPv6) EndpointSlices, but 0 mons is certainly
+		// an error state, so it's better to do nothing destructive right now.
+		logger.Debug("no mon addresses found, skipping endpointslice resource reconciliation")
+		return nil
+	}
+
+	ipv4Addresses := []string{}
+	ipv6Addresses := []string{}
+
+	for _, mon := range monitors {
+		host, _, err := net.SplitHostPort(mon.Endpoint)
+		if err != nil {
+			return errors.Wrapf(err, "failed to parse mon addr %q", mon.Endpoint)
+		}
+		ip := net.ParseIP(host)
+		if ip == nil {
+			logger.Warningf("invalid IP parsed from mon endpoint: %s", mon.Endpoint)
+			continue
+		}
+		if ip.To4() != nil {
+			ipv4Addresses = append(ipv4Addresses, host)
+		} else {
+			ipv6Addresses = append(ipv6Addresses, host)
+		}
+	}
+
+	if err := c.createEndpointSliceForAddresses(ipv4Addresses, discoveryv1.AddressTypeIPv4); err != nil {
+		return err
+	}
+
+	if err := c.createEndpointSliceForAddresses(ipv6Addresses, discoveryv1.AddressTypeIPv6); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Cluster) createEndpointSliceForAddresses(addresses []string, addressType discoveryv1.AddressType) error {
+	client := c.context.Clientset.DiscoveryV1().EndpointSlices(c.Namespace)
+
+	sliceName := endpointSliceNameIPv4
+	if addressType == discoveryv1.AddressTypeIPv6 {
+		sliceName = endpointSliceNameIPv6
+	}
+
+	if len(addresses) == 0 {
+		logger.Debugf("no %s addresses found, deleting existing %q endpointslice if exists", addressType, sliceName)
+		if err := client.Delete(c.ClusterInfo.Context, sliceName, metav1.DeleteOptions{}); err != nil {
+			if !kerrors.IsNotFound(err) {
```

Comment:
I'm currently working on improving the tests for this MR. My new tests cought the following bug:

```suggestion
			if kerrors.IsNotFound(err) {
```

Reply:
hmm whats your test procedure? I tried checking this case by creating empty endpointslice with different IPFamily and it worked.

Reply:
Actually, my tests didn't catch that bug, but I saw the bug in the logs, just like @travisn.

Reply:
ouch indeed, it should not be negated this way